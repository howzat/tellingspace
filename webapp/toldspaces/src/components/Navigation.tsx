import {Link} from "gatsby";
import React from "react";

export type NavigationProps = {
    links: Array<LinkData>;
}

export type LinkData = {
    text: string;
    to: string;
}

export const Navigation = ({links}: NavigationProps): JSX.Element => {
    console.log("SiteNavigation links", links)
    return (
      <header>
        <span data-testid="links">
      {
          links.map(v =>
            <div key={v.to} className={"container drop-shadow-md text-xs"}>
                <Link data-testid={`link-${v.to}`} key={v.to} to={v.to}
                      className={"box-decoration-slice bg-gradient-to-r from-teal-900 to-emerald-500 text-white p-1 m-1 rounded hover:bold"}>{v.text} and
                    all that
                </Link>
            </div>
          )
      }
        </span>
      </header>
    )
}

export const NewNavigationFromLocationsQuery = (data: Queries.LocationsJsonConnection): JSX.Element => {
    console.log('navigation items', data)
    return <Navigation data-testid="nav" links={
        data.nodes.map((node) => {
            return {to: node.sid, text: node.text}
        })
    }
    />
}
