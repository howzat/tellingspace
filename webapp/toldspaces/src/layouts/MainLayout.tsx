import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import CssBaseline from '@mui/material/CssBaseline';
import React from "react"
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import {NewNavigationFromLocationsQuery} from "../components/Navigation";
import {createTheme, ThemeProvider} from '@mui/material/styles';
import {IGatsbyImageData} from "gatsby-plugin-image/dist/src/components/gatsby-image.browser";
import {getImage} from "gatsby-plugin-image";
import {Copyright} from "../components/Copyright";
import {useAllLocations} from "../hooks/use-locations";
import {PageProps} from "gatsby";

export const theme = createTheme();

export type MainLayoutProps = {
    children?: React.ReactNode
}
const MainLayout = (props: PageProps<MainLayoutProps>) => {

    const locations: Array<Queries.LocationsJson> = useAllLocations()
    const imageMap = new Map<string, IGatsbyImageData>()
    locations.map((location) => {
        // @ts-ignore
        let i: IGatsbyImageData = getImage(location.image!);
        imageMap.set(location.sid, i)
    })

    return (
      <ThemeProvider theme={theme}>
          <CssBaseline/>
          <AppBar position="relative">
              <Toolbar>
                  <Typography variant="h6" component="div" color="inherit" noWrap
                              sx={{flexGrow: 1}}>TOLDSPACES</Typography>
                  {NewNavigationFromLocationsQuery()}
              </Toolbar>
          </AppBar>
          <main>
              <>
                  {props.children}
              </>
          </main>
          <Box sx={{bgcolor: 'background.paper', p: 6}} component="footer">
              <Typography variant="h6" align="center" gutterBottom>
                  Footer
              </Typography>
              <Typography
                variant="subtitle1"
                align="center"
                color="text.secondary"
                component="p"
              >
                  TOLDSPACES is an experiment in making things outside
              </Typography>
              <Copyright/>
          </Box>
      </ThemeProvider>

    )
}

export default MainLayout