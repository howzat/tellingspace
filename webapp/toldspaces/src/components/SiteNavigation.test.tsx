/**
 * @jest-environment jsdom
 */
import React from "react";
import {Navigation} from "./Navigation";
import TestRenderer from 'react-test-renderer';
import {useStaticQuery} from 'gatsby'
import {Link} from 'gatsby-theme-material-ui';
import {render} from "@testing-library/react";
import {MainLayout} from "../layouts/MainLayout";

beforeEach(() => {
    // @ts-ignore
    useStaticQuery.mockImplementationOnce(() => {
        return {
            allLocationsJson: {
                nodes: [
                    {
                        sid: "thethunderor",
                        text: "The Journey Begins and the Thunderor Awakes"
                    },
                    {
                        sid: "splinket",
                        text: "Location 2"
                    },
                    {
                        sid: "drblackpill",
                        text: "Location 3"
                    }
                ]
            }
        }
    })
})

describe("<Navigation />", () => {
    it("should render correctly", async () => {
        let testRenderer = TestRenderer.create(<Navigation
          links={[{to: "toValue1", text: "textValue1"}, {to: "toValue2", text: "textValue2"}]}/>);
        expect(testRenderer.toJSON()).toMatchSnapshot()
    });

    it("should display a list of links with the right 'to' fields", async () => {
        let testRenderer = TestRenderer.create(<Navigation
          links={[{to: "toValue1", text: "textValue1"}, {to: "toValue2", text: "textValue2"}]}/>);
        let root = testRenderer.root;

        let findAllByType = root.findAllByType(Link);
        expect(findAllByType[0].props.to).toBe("toValue1")
        expect(findAllByType[0].props.children).toContain("textValue1")

        expect(findAllByType[1].props.to).toBe("toValue2")
        expect(findAllByType[1].props.children).toContain("textValue2")
    });


    it("something about navs", async () => {
        let layout = render(<MainLayout/>);
        // screen.debug();
        let navigation = await layout.findByTestId("links");
        expect(navigation.children).toHaveLength(3);

        // const links = ([['thethunderor','thethunderor'], ['splinket', 'splinket'], ['drblackpill', 'drblackpill']]).map( ([toFind, text]) => {
        //     let navigationLink = layout.findAllByTestId(`link-${toFind}`);
        //     return (text, navigationLink)
        //     //???
        // }).then( p => {}
        //
        //
        // )
        //
        // const lll = await Promise.all(links).then(
        //   p =>
        //   console.log(p)
        // );
        //
        // console.log(lll)
    })
});