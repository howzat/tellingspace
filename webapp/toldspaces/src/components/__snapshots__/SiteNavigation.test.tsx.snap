// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> should render correctly 1`] = `
<nav
  className="flex items-center justify-between flex-wrap bg-teal-500 p-6"
>
  <div
    className="flex items-center flex-shrink-0 text-white mr-6"
  >
    <span
      className="font-semibold text-xl tracking-tight"
    >
      ToldSpaces
    </span>
  </div>
  <div
    className="block lg:hidden"
  >
    <button
      className="flex items-center px-3 py-2 border rounded text-teal-200 border-teal-400 hover:text-white hover:border-white"
    >
      <svg
        className="fill-current h-3 w-3"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Menu
        </title>
        <path
          d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z"
        />
      </svg>
    </button>
  </div>
  <div
    className="w-full block flex-grow lg:flex lg:items-center lg:w-auto"
  >
    <div
      className="text-sm lg:flex-grow"
    >
      <span
        data-testid="links"
      >
        <div
          className="container drop-shadow-md text-xs"
        >
          <a
            className="block mt-4 lg:inline-block lg:mt-0 text-teal-200 hover:text-white mr-4 rounded hover:bold"
            data-testid="link-toValue1"
            href="toValue1"
          >
            textValue1
          </a>
        </div>
        <div
          className="container drop-shadow-md text-xs"
        >
          <a
            className="block mt-4 lg:inline-block lg:mt-0 text-teal-200 hover:text-white mr-4 rounded hover:bold"
            data-testid="link-toValue2"
            href="toValue2"
          >
            textValue2
          </a>
        </div>
      </span>
    </div>
  </div>
</nav>
`;
